### PR TITLE
Improve test coverage for Python support

### PR DESCRIPTION
**Pull request by greenc-FNAL**
### Goal

The primary goal of this PR is to significantly improve the test coverage for the Phlex Python plugin and resolve several discovered issues related to memory management, type conversion, and core graph connectivity safety. It also modernizes the build system configuration and coverage tooling.

### Key Changes and Rationale

**1. Introduction of `phlex.Variant` Helper**
- **Changes**: Added a new Python class `Variant` in `plugins/python/python/phlex/__init__.py`.
- **Motivation**: Python algorithms are often generic, but Phlex registration requires specific C++ signatures. Additionally, Python parameter names in a function's `__annotations__` might not match the positional labels defined in the Phlex configuration.
- **Solution**: The `Variant` class allows developers to wrap a callable and associate it with custom annotations and a specific name. This enables the same Python function to be registered multiple times with different types without modifying the original function.
- **Rationale**: This decouples the algorithm implementation from the registration metadata. The C++ bridge was also updated to "unwrap" these objects, retrieving the underlying `phlex_callable` for performance while still using the provided annotations for type mapping.

**2. Robustness Improvements in `modulewrap.cpp`**
- **Reference Counting**: Added missing `Py_DECREF` calls in the error paths of `parse_args`, `md_transform`, and `md_observe`. This prevents Python object leaks when registration fails.
- **Sequence Handling**: Replaced manual sequence checks with `PySequence_Fast`. This allows Phlex to accept any Python sequence (lists, tuples, etc.) for input/output labels.
- **Type Matching & NumPy Fixes**: 
    - Corrected a typo where double-precision NumPy arrays were matched against `double64]]` instead of `float64]]`.
    - Updated type matching logic to use fixed-length substring comparisons, supporting PEP 604 union types (e.g., `ndarray | list`).
- **List-based Type Converters**: Added explicit support for `list[int]`, `list[float]`, and `list[double]`. 
    - **Rationale**: Provides a fallback for users not using NumPy or for cases where standard Python lists are more appropriate.

**3. Core Connectivity & Safety**
- **Changes**: Added null-pointer checks in `phlex/core/edge_maker.cpp` and `phlex/core/edge_maker.hpp`.
- **Motivation**: Misconfigurations in the Jsonnet graph could lead to null ports during edge creation, resulting in segmentation faults.
- **Solution**: The code now throws a `std::runtime_error` with a descriptive message identifying the offending nodes/ports.

**4. Build System & Tooling Modernization**
- **CMake Optimizations**:
    - Introduced `DART_TESTING_TIMEOUT` and `CTEST_TEST_TIMEOUT` to prevent long stalls in CI environments.
    - Refined GCC warning suppressions for versions 14+ to reduce noise from false-positive array-bounds and string-overflow warnings.
    - Standardized Sanatizer (ASan/TSan) compile options for better compatibility.
- **Coverage Tooling**: Updated `scripts/coverage.sh` and its documentation to support both Clang (LLVM source-based) and GCC (gcov) presets, enabling both high-fidelity local reports and CI-compatible XML outputs.
- **Documentation**: Added `plugins/python/README.md` (architecture guide) and `.github/copilot-instructions.md` (project context for AI assistants).

### New Tests

This PR adds a comprehensive suite of tests in `test/python/` to ensure long-term stability:

- **`unit_test_variant.py`**: Dedicated unit tests for the `Variant` class logic.
- **`test_types.py` & `test_coverage.py`**: Exercise various scalar, vector, and list-based type conversions to ensure data integrity.
- **`test_callbacks.py`**: Targets edge cases in `modulewrap.cpp`, such as 3-argument callbacks and Python exception propagation.
- **Error Path Verification**: Added several tests (`pybadbool.jsonnet`, `pybadint.jsonnet`, `pymismatch.jsonnet`, etc.) that verify Phlex correctly catches and reports type and signature mismatches.
- **Coverage Integration**: Updated `test/python/CMakeLists.txt` to support `pytest-cov`, enabling integrated Python coverage reporting.

### CODE REVIEW

**Comment by greenc-FNAL** [comment]
@phlexbot format

**Comment by greenc-FNAL** [comment]
@phlexbot python-fix

**Comment by greenc-FNAL** [comment]
@phlexbot python-fix

**Comment by wlav** [comment]
I wasn't aware of this PR when I started earlier this week on this one: https://github.com/Framework-R-D/phlex/pull/245 . On first glance, the two are mostly complementary and this one can benefit from the `Variant` class introduced in the other to simplify registration of the same Python function with different C++ types.

**Comment by greenc-FNAL** [comment]
@phlexbot format

**Comment by greenc-FNAL** [comment]
@phlexbot python-format

**Comment by greenc-FNAL** [comment]
@phlexbot python-fix

File: plugins/python/src/modulewrap.cpp [hunk 1]
```diff
@@ -343,34 +358,117 @@ namespace {
...
                                                                                                    \
-    return (intptr_t)pyll;                                                                         \
+    return PyObjectPtr((PyObject*)pyll, PyObjectDeleter());                                        \
   }
 
-  VECTOR_CONVERTER(vint, int, NPY_INT)
-  VECTOR_CONVERTER(vuint, unsigned int, NPY_UINT)
-  VECTOR_CONVERTER(vlong, long, NPY_LONG)
-  VECTOR_CONVERTER(vulong, unsigned long, NPY_ULONG)
+  // VECTOR_CONVERTER(vint, int, NPY_INT)
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: What is the motivation for expanding these? It leads to a lot of code duplication.

**Comment by greenc-FNAL** [comment]
@phlexbot format

**Comment by greenc-FNAL** [comment]
@phlexbot python-fix

**Comment by greenc-FNAL** [comment]
@phlexbot format

File: plugins/python/CMakeLists.txt
Comment (knoepfel) [inline plugins/python/CMakeLists.txt]: Looks like the changes in this file revert those made in #219.

**Review by knoepfel — COMMENTED** [review]
@greenc-FNAL, this looks like an improvement.  Thanks!  Some comments below for your consideration.  Most pressing is to make sure that the changes to the `CMakeLists.txt` files are intended and do not overwrite was changed with #219.

File: plugins/python/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+### 1. The "Type Bridge" (`modulewrap.cpp`)
+
+The core of the integration is the type conversion layer in `src/modulewrap.cpp`. This layer is responsible for:
+- Converting Phlex `Product` objects (C++) into Python objects (e.g., `PyObject*`, `numpy.ndarray`).
+- Converting Python return values back into Phlex `Product` objects.
+
+**Critical Implementation Detail:**
+The type mapping relies on **string comparison** of type names.
+- **Mechanism**: The C++ code checks `type_name() == "float64]]"` to identify a 2D array of doubles.
```
[the remark this answers is not in the document]
Comment (wlav) [reply plugins/python/README.md]: All of this array stuff is supposed to be temporary, given that support from the IDL should supersede it. If, however, this is supposed to become a permanent feature then, now that we have NumPy as a proper dependency, I'd just pull the names from `numpy.typing`.

File: CMakeLists.txt [hunk 3]
```diff
@@ -70,13 +70,10 @@ add_compile_options(
...
-  if(
-    CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "14.1"
-    AND CMAKE_COMPILER_VERSION VERSION_LESS "15"
-  )
-    # GCC 14.1 issues many false positives re. array-bounds and
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "14.1")
+    # GCC 14.1+ issues many false positives re. array-bounds and
     # stringop-overflow
-    add_compile_options(-Wno-array-bounds -Wno-stringop-overflow)
+    add_compile_options(-Wno-array-bounds -Wno-stringop-overflow -Wno-maybe-uninitialized)
```
[the remark this answers is not in the document]
Comment (greenc-FNAL) [reply CMakeLists.txt]: I thought so too, but I don't think I intentionally added this back in. Will try again with it out.

File: test/demo-giantdata/unfold_transform_fold.cpp [hunk 4]
```diff
@@ -36,7 +36,7 @@ TEST_CASE("Unfold-transform-fold pipeline", "[concurrency][unfold][fold]")
   // Test parameters - moderate scale to ensure sustained concurrent execution
   constexpr std::size_t n_runs = 1;
   constexpr std::size_t n_subruns = 1;
-  constexpr std::size_t n_spills = 20;
+  constexpr std::size_t n_spills = 100;
```
[the remark this answers is not in the document]
Comment (greenc-FNAL) [reply test/demo-giantdata/unfold_transform_fold.cpp]: AI shenanigans, AFAICT.

**Comment by greenc-FNAL** [comment]
@phlexbot python-fix

**Comment by greenc-FNAL** [comment]
@phlexbot format

File: plugins/python/src/modulewrap.cpp [hunk 5]
```diff
@@ -292,17 +315,35 @@ namespace {
...
   }                                                                                                \
                                                                                                    \
-  static cpptype py_to_##name(intptr_t pyobj)                                                      \
+  static cpptype py_to_##name(PyObjectPtr pyobj)                                                   \
   {                                                                                                \
     PyGILRAII gil;                                                                                 \
-    cpptype i = (cpptype)frompy((PyObject*)pyobj);                                                 \
-    Py_DECREF((PyObject*)pyobj);                                                                   \
+    cpptype i = (cpptype)frompy(pyobj.get());                                                      \
+    if (PyErr_Occurred()) {                                                                        \
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: It'll be easier to use `msg_from_py_error` (from `errorwrap.hpp`) to make sure all python errors are consistently treated wrt. their propagation through the Phlex C++ code and creation of Python traceback.

**Comment by greenc-FNAL** [comment]
@phlexbot jsonnet-format-fix

**Comment by greenc-FNAL** [comment]
@phlexbot format

**Comment by greenc-FNAL** [comment]
@phlexbot format

File: plugins/python/src/modulewrap.cpp [hunk 6]
```diff
@@ -57,35 +57,46 @@ namespace {
...
     }
-    return (PyObject*)arg;
+    return pyobj;
   }
 
   // callable object managing the callback
   template <size_t N>
   struct py_callback {
-    PyObject const* m_callable; // owned
+    PyObject* m_callable; // owned
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: These were `const` b/c the phlex interface required it. Fine if that's no longer true.

File: plugins/python/src/modulewrap.cpp [hunk 7]
```diff
@@ -57,35 +57,46 @@ namespace {
...
     {
-      Py_INCREF(callable);
+      Py_XINCREF(callable);
       m_callable = callable;
     }
     py_callback(py_callback const& pc)
     {
-      Py_INCREF(pc.m_callable);
+      PyGILRAII gil;
+      Py_XINCREF(pc.m_callable);
```
[the remark this answers is not in the document]
Comment (wlav) [reply plugins/python/src/modulewrap.cpp]: Meaning, every use that I can think of has the GIL already held?

File: plugins/python/src/modulewrap.cpp [hunk 8]
```diff
@@ -94,8 +105,23 @@ namespace {
 
       PyGILRAII gil;
 
-      PyObject* result =
-        PyObject_CallFunctionObjArgs((PyObject*)m_callable, lifeline_transform(args)..., nullptr);
+      PyObject* arg_tuple = PyTuple_New(N);
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: What is the reason/motivation/benefit of this change?

File: plugins/python/src/modulewrap.cpp [hunk 9]
```diff
@@ -130,16 +172,17 @@ namespace {
...
+      }
     }
 
   private:
     template <typename... Args>
     void decref_all(Args... args)
     {
       // helper to decrement reference counts of N arguments
-      (Py_DECREF((PyObject*)args), ...);
+      (Py_XDECREF((PyObject*)args), ...);
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: Same as above. Args aren't zero here or the code would have seg-faulted already in the python call.

File: plugins/python/src/modulewrap.cpp [hunk 10]
```diff
@@ -207,28 +250,42 @@ namespace {
     std::string ann;
     if (!PyUnicode_Check(pyobj)) {
       PyObject* pystr = PyObject_GetAttrString(pyobj, "__name__"); // eg. for classes
+
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: More likely an error message is then due, as there's no guarantee that we know how to map a generic (or a protocol) to a C++ type. Some exceptions, such as sequences, may exist.

File: plugins/python/src/modulewrap.cpp [hunk 11]
```diff
@@ -279,7 +336,7 @@ namespace {
     unsigned long ul = PyLong_AsUnsignedLong(pyobject);
     if (ul == (unsigned long)-1 && PyErr_Occurred() && PyLong_Check(pyobject)) {
       PyErr_Clear();
-      long i = PyLong_AS_LONG(pyobject);
+      long i = PyLong_AsLong(pyobject);
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: Unnecessary since we just did `PyLong_Check`.

File: plugins/python/src/modulewrap.cpp [hunk 12]
```diff
@@ -486,15 +797,13 @@ static PyObject* parse_args(PyObject* args,
...
       output_types.push_back(annotation_as_text(ret));
 
-    // dictionary is ordered with return last if provide (note: the keys here
-    // could be used as input labels, instead of the ones from the configuration,
-    // but that is probably not practical in actual use, so they are ignored)
-    PyObject* values = PyDict_Values(annot);
-    for (Py_ssize_t i = 0; i < (PyList_GET_SIZE(values) - (ret ? 1 : 0)); ++i) {
-      PyObject* item = PyList_GET_ITEM(values, i);
-      input_types.push_back(annotation_as_text(item));
+    Py_ssize_t pos = 0;
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: This still assumes ordering and if ordered, the original is fine.

**Comment by greenc-FNAL** [comment]
Review the [full CodeQL report](https://github.com/Framework-R-D/phlex/security/code-scanning) for details.

**Comment by greenc-FNAL** [comment]
Review the [full CodeQL report](https://github.com/Framework-R-D/phlex/security/code-scanning) for details.

**Comment by greenc-FNAL** [comment]
@phlexbot format

File: plugins/python/src/modulewrap.cpp [hunk 13]
```diff
@@ -130,16 +181,17 @@ namespace {
...
+      }
     }
 
   private:
     template <typename... Args>
     void decref_all(Args... args)
     {
       // helper to decrement reference counts of N arguments
-      (Py_DECREF((PyObject*)args), ...);
+      (Py_XDECREF((PyObject*)args), ...);
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: args can't be null as the conversion node will have thrown

File: plugins/python/src/modulewrap.cpp [hunk 14]
```diff
@@ -279,7 +345,7 @@ namespace {
     unsigned long ul = PyLong_AsUnsignedLong(pyobject);
     if (ul == (unsigned long)-1 && PyErr_Occurred() && PyLong_Check(pyobject)) {
       PyErr_Clear();
-      long i = PyLong_AS_LONG(pyobject);
+      long i = PyLong_AsLong(pyobject);
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: Not needed b/c above is a `PyLong_Check()` already.

File: plugins/python/src/modulewrap.cpp [hunk 15]
```diff
@@ -452,8 +539,13 @@ static PyObject* parse_args(PyObject* args,
...
   }
 
-  if (!PySequence_Check(input) || (output && !PySequence_Check(output))) {
-    PyErr_SetString(PyExc_TypeError, "input and output need to be sequences");
+  if (!PyList_Check(input) && !PyTuple_Check(input)) {
+    PyErr_SetString(PyExc_TypeError, "input parameter must be a list or tuple");
+    return nullptr;
+  }
+
+  if (output && !PyList_Check(output) && !PyTuple_Check(output)) {
```
Comment (wlav) [inline plugins/python/src/modulewrap.cpp]: Same as above? Any object implementing the sequence protocol should do?

File: plugins/python/src/modulewrap.cpp [hunk 16]
```diff
@@ -57,35 +57,51 @@ namespace {
...
       }
+      return *this;
+    }
+    ~py_callback()
+    {
+      // Check if Python is still initialized before attempting cleanup
+      // While this check may not be reliable in all threading scenarios (e.g., offloaded threads),
+      // it prevents crashes during normal interpreter shutdown
+      if (Py_IsInitialized()) {
+        PyGILRAII gil;
```
[the remark this answers is not in the document]
Comment (greenc-FNAL) [reply plugins/python/src/modulewrap.cpp]: What happens if you try to grab the GIL and we're no longer initialized due to shutdown?

File: plugins/python/src/modulewrap.cpp [hunk 17]
```diff
@@ -479,24 +571,45 @@ static PyObject* parse_args(PyObject* args,
       Py_DECREF(callm);
     }
   }
-  Py_DECREF(sann);
```
[the remark this answers is not in the document]
Comment (greenc-FNAL) [reply plugins/python/src/modulewrap.cpp]: You're right, no reason to delay.

File: plugins/python/src/modulewrap.cpp [hunk 18]
```diff
@@ -353,44 +431,53 @@ namespace {
...
-    vec->insert(vec->end(), raw, raw + total);                                                     \
+      /* copy the array info; note that this assumes C continuity */                               \
+      cpptype* raw = static_cast<cpptype*>(PyArray_DATA(arr));                                     \
+      vec->reserve(total);                                                                         \
+      vec->insert(vec->end(), raw, raw + total);                                                   \
+    } else if (pyobj && PyList_Check((PyObject*)pyobj)) {                                          \
+      Py_ssize_t total = PyList_Size((PyObject*)pyobj);                                            \
+      vec->reserve(total);                                                                         \
+      for (Py_ssize_t i = 0; i < total; ++i) {                                                     \
+        PyObject* item = PyList_GetItem((PyObject*)pyobj, i);                                      \
```
[the remark this answers is not in the document]
Comment (wlav) [reply plugins/python/src/modulewrap.cpp]: It may also simplify things to use sequences anyway. The copying of the array isn't desirable either, but is what is is for now, so could pass everything through the sequence protocol until we have IDL support.

File: plugins/python/src/modulewrap.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (wlav) [reply plugins/python/src/modulewrap.cpp]: Well, until we get a proper place to finalize the interpreter (see other discussion), `atexit()` will not be called either... As for `valgrind`, this being a pure Python object, its memory sits in an arena that should be de-allocated wholesale. It would not surprise me if `valgrind` is happy.

**Comment by greenc-FNAL** [comment]
@phlexbot clang-fix

**Comment by greenc-FNAL** [comment]
Review the [full CodeQL report](https://github.com/Framework-R-D/phlex/security/code-scanning) for details.

**Review by wlav — APPROVED** [review]
I think I've given all the relevant feedback I can at this point. :)

**Comment by greenc-FNAL** [comment]
Review the [full CodeQL report](https://github.com/Framework-R-D/phlex/security/code-scanning) for details.
